extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/ImportFrameHandler.scala

```diff
@@ -41,28 +42,83 @@ class ImportFrameHandler extends Handler {
     val collectDomainsTask = new CollectCategoricalDomainsTask(frameKey)
     collectDomainsTask.doAllNodes()
     val stringDomains = collectDomainsTask.getDomains()
-    val domains = expandDomains(stringDomains, columnTypes)
+    val isEmpty = stringDomains == null
+    val columnTypesAwareOfEmptyFrames = if (isEmpty) {
+      columnTypes.map(columnType => if (columnType == Vec.T_CAT) Vec.T_STR else columnType)
+    } else {
+      columnTypes
+    }
+    val domains = expandDomains(stringDomains, columnTypesAwareOfEmptyFrames)
+    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypesAwareOfEmptyFrames, domains)
+    val frame = DKV.getGet[Frame](frameKey)
 
-    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypes, domains)
+    convertColumnsWithTooManyCategoricalLevelsToStringColumns(frame, columnTypesAwareOfEmptyFrames, stringDomains)
 
-    // Update categorical indices for all chunks.
-    val frame = DKV.getGet[Frame](frameKey)
     val categoricalColumnIndices = for ((vec, idx) <- frame.vecs().zipWithIndex if vec.isCategorical) yield idx
+
+    // Update categorical indices for all chunks. (local chunk domains -> global domains)
     val updateCategoricalIndicesTask = new UpdateCategoricalIndicesTask(frameKey, categoricalColumnIndices)
     updateCategoricalIndicesTask.doAll(frame)
 
-    // Convert categorical columns with too many categories to T_STR
-    categoricalColumnIndices.foreach { catColIdx =>
-      val vector = frame.vec(catColIdx)
-      if (vector.cardinality() > Categorical.MAX_CATEGORICAL_COUNT) {
-        Log.warn(
-          s"Categorical column with index '$catColIdx' exceeded maximum number of categories. " +
+    // Convert unique categorical columns to T_STR
+    convertUniqueCategoricalColumnsToStringColumns(frame, categoricalColumnIndices)
```

Comment:
I think it might be a good idea to keep this logic on the H2O side - close to the PreviewParseWriter - then we can make sure the logic is actually consistent.

Reply:
I can add the cases also for T_UUID and T_TIME. But do you mean to add a piece of code to H2O-3 repository which will utilized here?

Reply:
the decission logic is done via `PreviewParseWriter`, but without `T_UUID` and `T_TIME`